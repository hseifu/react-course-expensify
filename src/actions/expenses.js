import uuid from 'uuid';

//ADD EXPENSE
export const addExpense = (
    { 
        description = '',
        note = '', 
        amount = 0, 
        createdAt = 0
    } = {}) => ({
    type: 'ADD_EXPENSE',
    expense: {
        id: uuid(),
        description,
        note,
        amount,
        createdAt
    }
})

//Remove Expense
export const removeExpense = (removeId) => ({
    type: 'REMOVE_EXPENSE',
    removeId
})

//Edit Expense
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})