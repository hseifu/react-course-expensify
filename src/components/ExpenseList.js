import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from '../components/ExpenseListItem';
import getVisibleExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-wide">Expense</div>
            <div className="show-for-wide">Amount</div>
        </div>
        <div className="list-body">
            {
                props.expenses.length == 0 ? <div className="list-item list-item--message"><span> No expenses </span></div> :
                props.expenses.map((expense) => {
                    return (
                    <ExpenseListItem 
                        key={expense.id} 
                        {...expense}
                    />)
                })
            }
        </div>
        
    </div>
)

const mapStateToProps = (state) => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseList);