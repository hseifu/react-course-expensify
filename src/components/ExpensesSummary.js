import React from 'react';
import { connect } from 'react-redux';
import getVisibleExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpensesSummary = (props) => {
    return (
        <div>
            <h4>Viewing {props.expenses.length} expense{props.expenses.length == 1 ? '' : 's'} totalling {numeral(getExpensesTotal(props.expenses)/100).format('$0,0.00')}</h4>
        </div>
    )
}

const mapStateToProps = (state) => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpensesSummary);