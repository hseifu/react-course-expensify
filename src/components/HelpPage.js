import React from 'react';

const HelpPage = () => (
    <div>
        this is my Help page
    </div>
)

export default HelpPage;