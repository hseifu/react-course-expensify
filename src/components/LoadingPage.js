import React from 'react';

const LoadingPage = () => (
    <div className="loader">
        <img className="loader__img" src="/images/loading.gif"/>
    </div>
);

export default LoadingPage;