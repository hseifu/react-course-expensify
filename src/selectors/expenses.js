import moment from 'moment';

//get visible data
export default (expenses, {text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment,'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment,'day') : true;
        const reg = new RegExp('^'+text.toLowerCase()+'|\\s'+text.toLowerCase(),'g');
        const textMatch = expense.description.toLowerCase().match(reg);
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a,b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1;
        }
        else {
            return a.amount < b.amount ? 1 : -1
        }
    })
}