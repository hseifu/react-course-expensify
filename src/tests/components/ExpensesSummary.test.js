import React from 'react';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';

test("should setup 1 expense summary", () => {
    const wrapper = shallow(<ExpensesSummary expenses={[expenses[0]]}/>);
    expect(wrapper).toMatchSnapshot();
})

test("should setup 1 expense summary", () => {
    const wrapper = shallow(<ExpensesSummary expenses={expenses}/>);
    expect(wrapper).toMatchSnapshot();
})