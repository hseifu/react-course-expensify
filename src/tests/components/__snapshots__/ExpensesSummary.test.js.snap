// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should setup 1 expense summary 1`] = `
<div>
  <h4>
    Viewing 
    1
     expense
     totalling 
    $1.95
  </h4>
</div>
`;

exports[`should setup 1 expense summary 2`] = `
<div>
  <h4>
    Viewing 
    3
     expense
    s
     totalling 
    $19,951.95
  </h4>
</div>
`;
